File: DP/91.decode_ways.py
# Decode 1->A, 26->Z
# How many ways to decode a string containing only integers?

class Solution:
    def oneNum(self, s: str) -> int:
        return int(int(s) != 0)
    
    def twoNums(self, s: str) -> int:
        return int(int(s) >= 10 and int(s) <= 26)
    
    def numDecodings(self, s: str) -> int:
        n = len(s)
        if n == 1:
            return self.oneNum(s)
        elif n == 2:
            return self.twoNums(s) + int((self.oneNum(s[0]) and self.oneNum(s[1])))
        else:
            dp = []
            dp.append(self.oneNum(s[0]))
            dp.append(self.twoNums(s[0:2]) + int(self.oneNum(s[0]) and self.oneNum(s[1])))
            for i in range(2, n):
                dp.append(self.oneNum(s[i]) * dp[i-1] + self.twoNums(s[i-1:i+1]) * dp[i-2])
            return dp[-1]
        
solution = Solution()
print(solution.numDecodings("189327"))

File: C++/Dynamic Programming/91. Decode Ways.cpp
class Solution {
public:
    
    int numDecodings(string s) {
        int m1 = 1, m2 = s[0]=='0' ? 0 : 1;
        if (s.size() == 1) return m2;
        for (int i = 1; i < s.size(); i++) {
            int temp = m2;
            string subs = s.substr(i-1, 2);
            m2 = m1 * ((subs >= "10" && subs <= "26") ? 1 : 0) + m2 * (s[i] == '0' ? 0 : 1);
            m1 = temp;
        }
        return m2;
    }
};